flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java

```diff
@@ -186,7 +186,7 @@ public void abortCheckpointOnBarrier(long checkpointId, Throwable cause, Operato
 
 		checkpointStorage.clearCacheFor(checkpointId);
 
-		channelStateWriter.abort(checkpointId, cause);
+		channelStateWriter.abort(checkpointId, cause, true);
```

Comment:
I am not quite sure whether we still have race condition here. `abortCheckpointOnBarrier` might be called from `CheckpointBarrierUnaligner` after triggering checkpoint into mailbox. After aborting, we did not remove the checkpoint action from mailbox, so the checkpoint might still happen afterwards. So how to guarantee that `#getAndRemoveWriteResult` would never be called after aborting?

Reply:
I got the answer when seeing the following commit "Ignore out of order checkpoints in SubtaskCheckpointCoordinator". It updates the `lastCheckpointId` to prevent the following aborted checkpoint execution.

Reply:
Right, there are several race conditions that are addressed in different commits.